### PR TITLE
Disable leading zero stripping by default

### LINKED ISSUE #12870: [BUG] DatetimeTickFormatter strips leading 0 in "%H" format, works in " %H"
### Software versions

Python version      :  3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)]
IPython version     :  8.10.0
Tornado version     :  6.2
Bokeh version       :  3.0.3
BokehJS static path :  .venv\lib\site-packages\bokeh\server\static
node.js version     :  v18.14.1
npm version         :  9.3.1
Operating system    :  Windows-10-10.0.19045-SP0

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

"%H" always has 2 digits

### Observed behavior

"%H" doesn't always have 2 digits

### Example code

```Python
from datetime import datetime, timedelta

from bokeh.io import show
from bokeh.plotting import figure
from bokeh.models.formatters import DatetimeTickFormatter

x = [datetime(2023, 1, 1) + timedelta(hours=h) for h in range(25)]
y = list(range(len(x)))

p = figure(x_axis_type="datetime")
p.line(x, y)
p.xaxis.formatter = DatetimeTickFormatter(hours="%H")  # " %H" works
show(p)
```

### Stack traceback or browser console output

_No response_

### Screenshots

`DatetimeTickFormatter(hours="%H")`

`DatetimeTickFormatter(hours=" %H")` (notice leading space in format string)

### PR DESCRIPTION
**Pull request by muendlein**
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [x] issues: fixes #12870 
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

### CODE REVIEW

File: src/bokeh/models/formatters.py [hunk 1]
```diff
@@ -636,7 +637,15 @@ def __init__(self, *args, **kwargs) -> None:
...
-    strip_leading_zeros = Bool(default=True, help="Whether to strip any leading zeros in the formatted ticks")
+    strip_leading_zeros = Either(Bool, List(Enum(ResolutionType)), default=False,
+                                 help="""
+                                 Whether to strip any leading zeros in the formatted ticks.
+                                 Valid values are:
+
+                                 * True or False (default) to set stripping across all resolutions.
+                                 * A list of resolution types, e.g. `["microseconds", "milliseconds"]`, to enable
+                                   scale-dependent stripping of leading zeros.
+                                 """)
```
Comment (mattpap) [inline src/bokeh/models/formatters.py]:
```suggestion
    strip_leading_zeros = Either(Bool, List(Enum(ResolutionType)), default=False, help="""
    Whether to strip any leading zeros in the formatted ticks.
        
    Valid values are:
    * ``True`` or ``False`` (default) to set stripping across all resolutions.
    * A list of resolution types, e.g. ``["microseconds", "milliseconds"]``, to enable
    scale-dependent stripping of leading zeros.
    """)
```

File: bokehjs/src/lib/models/formatters/datetime_tick_formatter.ts [hunk 2]
```diff
@@ -167,7 +165,7 @@ export class DatetimeTickFormatter extends TickFormatter {
     let s = s0
     let hybrid_handled = false
     let next_index = resolution_index
-    let next_resolution: ResolutionType
+    let next_resolution: ResolutionType = resolution
```
Comment (mattpap) [inline bokehjs/src/lib/models/formatters/datetime_tick_formatter.ts]:
```suggestion
    let next_resolution = resolution
```

File: bokehjs/src/lib/models/formatters/datetime_tick_formatter.ts [hunk 3]
```diff
@@ -196,9 +194,10 @@ export class DatetimeTickFormatter extends TickFormatter {
       s = _strftime(t, this[next_resolution])
     }
 
-    if (this.strip_leading_zeros) {
+    const final_resolution = next_resolution
+    if ((typeof this.strip_leading_zeros === 'boolean' && this.strip_leading_zeros) || (Array.isArray(this.strip_leading_zeros) && this.strip_leading_zeros.includes(final_resolution))) {
```
Comment (mattpap) [inline bokehjs/src/lib/models/formatters/datetime_tick_formatter.ts]:
```suggestion
    const {strip_leading_zeros} = this
    if ((isBoolean(strip_leading_zeros) && strip_leading_zeros) || 
        (isArray(strip_leading_zeros) && strip_leading_zeros.includes(final_resolution))) {
```
`isBoolean` and `isArray` come from `core/util/types` module.

**Comment by mattpap** [comment]
Among other CI failures to fix, run `pytest tests/test_defaults.py` and then add and commit the changes to `tests/baselines/defaults.json5` and re-run the test to verify. Changes must be committed, otherwise you'll get failures again.

**Comment by muendlein** [comment]
@mattpap Thanks for the review/guidance! As for the final task  (_release document entry (if new feature or API change)_),  what exactly needs to be done in this case (API extension)?

**Comment by mattpap** [comment]
> Thanks for the review/guidance! As for the final task (release document entry (if new feature or API change)), what exactly needs to be done in this case (API extension)?

Given you've already updated the relevant docstring, I think it sufficiently covers this task. We may need an entry in https://github.com/bokeh/bokeh/wiki/Migration-Guides though, given the change in the default value.

File: src/bokeh/models/formatters.py [hunk 4]
```diff
@@ -636,7 +637,14 @@ def __init__(self, *args, **kwargs) -> None:
...
                    default="%Y").accepts(List(String), _deprecated_datetime_list_format)
 
-    strip_leading_zeros = Bool(default=True, help="Whether to strip any leading zeros in the formatted ticks")
+    strip_leading_zeros = Either(Bool, List(Enum(ResolutionType)), default=False, help="""
+    Whether to strip any leading zeros in the formatted ticks.
+
+    Valid values are:
+    * ``True`` or ``False`` (default) to set stripping across all resolutions.
+    * A list of resolution types, e.g. ``["microseconds", "milliseconds"]``, to enable
+    scale-dependent stripping of leading zeros.
```
Comment (bryevdv) [inline src/bokeh/models/formatters.py:647]: Would `Sequence` property work here? If I was personally writing code, I would aways reach for tuples before lists.

Comment (muendlein) [reply src/bokeh/models/formatters.py:647]: @bryevdv Sequence property should be doable here as mutability is not required. I will try to provide a commit for that the next days.

**Review by mattpap — APPROVED** [review]

**Comment by mattpap** [comment]
@muendlein, thanks.

**Pull request merged**